specs/extension_image.md

```diff
@@ -34,9 +34,103 @@ sysext images extend `/usr/` (OS vendor tree) and/or `/opt/` (third-party vendor
+2. *Mutable mode* - Writes are directed to an _upperdir_ specified by the user / operator
+    (see "Mutability Mode Configuration" below).
+    This _upperdir_ will contain all changes made to the overlaid base directory.
+    1. _Upperdir_ may be specified to _be_ the base directory: may be used to retain mutability
+       of the base directory after extensions have been merged.
+    2. Alternatively, _upperdir_ may be an entirely separate directory: modifications will be captured
+       but the base directory will remain unchanged, retaining its state from before the extension was merged.
+3. *Ephemeral Mode* - Similar to mutable mode (2.) above but writes are only stored temporarily while
+    extensions are merged, and discarded as soon as extensions are un-merged.
+    Location of temporary storage is implementation specific.
+    Useful for e.g. development and for one-shot validation operations.
+
+### Mutability Mode Configuration
+Immutable mode is the default.
+If none of the configurations outlined below were specified then extension overlays operate
+in immutable mode and base directories are read-only.
+Implementation of any of the below mutable configurations is optional.
+_If_ mutable modes are supported by an implementation, configuration option 1. below _must_ be supported
+for compatibility across implementations.
+
+Mutability modes may be configured in the following ways:
+1. By creating qualified paths or soft-links below `/var/lib/extension-data/`.
+   See "Qualified Paths Definition" below for details.
+   This is the most portable option across different implementations.
+   If an implementation supports base directory mutability then this mode _must_ be supported.
+2. By setting a respective option in an implementation's configuration file.
+   This option is implementation specific.
+   Implementations may choose to support a single option, multiple options for
+   system and configuration extensions, and/or multiple options per base directory.
+   Using this option should selectively override any qualified path definitions from 1.
+3. By passing a command line parameter upon extension merge or refresh.
+   This option is implementation specific.
+   Implementations may choose to support a single option and/or multiple options per base directory.
+   Using this option should selectively override any configurations from 1. and 2.
+
+#### Qualified Paths Definition
+
+Mutability Mode 1 enables mutability by creating paths or soft-links below
+`/var/lib/extension-data/`.
+Qualified paths are:
+* `/var/lib/extension-data/etc.local` - directory or soft-link to a directory to store writes to
+  `/etc`. This is for configuration extensions.
+* `/var/lib/extension-data/usr.local` - directory or soft-link to a directory to store writes to
+  `/usr`. This is for system extensions.
+* `/var/lib/extension-data/opt.local` - directory or soft-link to a directory to store writes to
+  `/opt`. This is for system extensions.
```

Comment:
Well, with the discussion above I am pretty sure that mutable extensions should look exactly like regular extensions, so that we can turn them into them. Or in other words, the stuff that shows up under /etc/ should be in a dir called "etc", not under one called "etc.local".

Or in other words:

/var/lib/sysexts.mutable/opt/
/var/lib/sysexts.mutable/usr/
/var/lib/confexts.mutable/etc/

are the actually populated dirs?

Reply:
I never liked the .local suffix either, so that looks better

Reply:
Good point, will update using Kai's proposal.